build.clj

```diff
@@ -0,0 +1,147 @@
+
+  (println "Building submodule " mod-name)
+  (let [basis (b/create-basis {:project "deps.edn"
+                               :aliases [:overrides]
+                               :extra {:aliases {:overrides
+                                                 {:override-deps
+                                                  {'io.github.swirrl/grafter.core {:mvn/version version}
+                                                   'io.github.swirrl/grafter.io {:mvn/version version}
+                                                   'io.github.swirrl/grafter.repository {:mvn/version version}}}}}})
+
+        mod-target (str (io/file root-target mod-name))
+        jar-file (str (io/file root-target jar-file))]
+
+    (b/write-pom (assoc opts
+                        :basis basis
+                        :class-dir mod-target))
+
+    (b/copy-dir {:src-dirs   (:paths basis)
+                 :target-dir mod-target})
+
+    (println "Building jar" (str jar-file "..."))
+
+    (b/jar (assoc opts
+                  :class-dir mod-target
+                  :jar-file jar-file))
+
+    ;;(println "Installing " jar-file " " mod-target)
+    (b/install (assoc opts
+                      :basis basis
+                      :jar-file jar-file
+                      :class-dir mod-target
+                      ))))
+
+(def submodules ["grafter.core" "grafter.io" "grafter.repository"])
+
+(defn- canonicalise-file [& f]
+  (str (apply io/file (.getCanonicalFile (io/file (or b/*project-root* ".")))
+              f)))
+
+(defn build-grafter-repo [opts]
+  (with-project-root "grafter.repository"
+    (b/process {:command-args ["clojure" "-T:build" "compile-java"]})))
+
+(defn clean-all [opts]
+  (b/delete {:path "./target"})
+  (b/delete {:path "./grafter.core/target"})
+  (b/delete {:path "./grafter.repository/target"}))
+
+(defn build-all-modules [opts]
+  (let [root-target (canonicalise-file "target")]
+    (doseq [module submodules]
+      (with-project-root module
+        (build-submodule :mod-name module
+                         :jar-file (format "%s-%s.jar" module version)
+                         :src-pom (canonicalise-file "template" (str module "-pom.xml"))
+                         :module-target (canonicalise-file module "target")
+                         :lib (symbol "io.github.swirrl" module)
+                         :version version
+                         :root-target root-target)))))
+
+(defn build-all
+  "Build all submodules locally"
+  [opts]
+  ;; first prep this dep by building its java classes
+  (aborting-process {:command-args ["clojure" "-X:deps" "prep"]}) ;; TODO is this necessary?
+
+  (clean-all opts)
+  (build-grafter-repo opts)
+  (test-all opts)
+
+  (build-all-modules opts))
+
+
+(defn deploy-all [opts]
+  (doseq [module submodules]
+    (dd/deploy {:artifact (format "./target/%s-%s.jar" module version)
+                :installer :remote
+                :sign-releases? false ;; TODO
```

Comment:
Flagging TODO in case you meant to come back to it before this got merged. (unclear to me how important it is, quick google doesn't leave me much the wiser 🤔)

Reply:
Good question; I put a TODO here to double check the status of signed-releases and verification, as `sign-releases? true` will require `gpg` and keys setup in CI too, and I wanted to double check whether it was worth doing.

Thanks to you spotting this I have now double checked, and it's basically not worth it, because there's currently no way to automatically verify web-of-trust signatures.

See discussion here: 

https://clojurians.slack.com/archives/C0H28NMAS/p1675338310363909